## Deep Learning with Keras: Assignment (PART D)

In [1]:
# import keras libraries
import numpy as np
import pandas as pd

In [2]:
# load concrete dataset
concrete = pd.read_csv("concrete_data.csv")
concrete.head()

Cement  Blast Furnace Slag  Fly Ash  Water  Superplasticizer  \
0   540.0                 0.0      0.0  162.0               2.5   
1   540.0                 0.0      0.0  162.0               2.5   
2   332.5               142.5      0.0  228.0               0.0   
3   332.5               142.5      0.0  228.0               0.0   
4   198.6               132.4      0.0  192.0               0.0   

   Coarse Aggregate  Fine Aggregate  Age  Strength  
0            1040.0           676.0   28     79.99  
1            1055.0           676.0   28     61.89  
2             932.0           594.0  270     40.27  
3             932.0           594.0  365     41.05  
4             978.4           825.5  360     44.30

In [3]:
# identify concrete dataset
concrete.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1030 entries, 0 to 1029
Data columns (total 9 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   Cement              1030 non-null   float64
 1   Blast Furnace Slag  1030 non-null   float64
 2   Fly Ash             1030 non-null   float64
 3   Water               1030 non-null   float64
 4   Superplasticizer    1030 non-null   float64
 5   Coarse Aggregate    1030 non-null   float64
 6   Fine Aggregate      1030 non-null   float64
 7   Age                 1030 non-null   int64  
 8   Strength            1030 non-null   float64
dtypes: float64(8), int64(1)
memory usage: 72.6 KB


### D. Increase the number of hidden layers

In [4]:
# import libraries for buidling neural nentwork model
import keras
from keras.models import Sequential
from keras.layers import Dense

In [5]:
# create a deep learning model
model = Sequential()
ncols = concrete.drop(columns=["Strength"], axis=1).shape[1]

# add hidden layers
# create optimizer
model.add(Dense(10, activation="relu", input_shape = (ncols,)))
model.add(Dense(10, activation="relu"))
model.add(Dense(10, activation="relu"))
model.add(Dense(1))
model.compile(optimizer="adam", loss="mean_squared_error")

d:\User\LocalMachine\Anaconda\Lib\site-packages\keras\src\layers\core\dense.py:87: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


### Model Training

In [6]:
# import scikit learn library for training and testing data
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_squared_error 

In [7]:
concrete.head()

Cement  Blast Furnace Slag  Fly Ash  Water  Superplasticizer  \
0   540.0                 0.0      0.0  162.0               2.5   
1   540.0                 0.0      0.0  162.0               2.5   
2   332.5               142.5      0.0  228.0               0.0   
3   332.5               142.5      0.0  228.0               0.0   
4   198.6               132.4      0.0  192.0               0.0   

   Coarse Aggregate  Fine Aggregate  Age  Strength  
0            1040.0           676.0   28     79.99  
1            1055.0           676.0   28     61.89  
2             932.0           594.0  270     40.27  
3             932.0           594.0  365     41.05  
4             978.4           825.5  360     44.30

In [8]:
# normalise data
def normalise_data(data, feature):
  # detertmine the mean value and std from input feature
  x_mean = data[feature].mean()
  x_std = data[feature].std()

  # apply z score to normalise input feature's data
  x_norm = (data[feature] - x_mean)/x_std
  return x_norm

def apply_data_normalisation(data, feature_list):
  norm_data = data
  for feature in feature_list:
    feat_norm = normalise_data(data, feature)
    norm_data.loc[:,feature] = feat_norm
  return norm_data

In [9]:
concrete.head()

Cement  Blast Furnace Slag  Fly Ash  Water  Superplasticizer  \
0   540.0                 0.0      0.0  162.0               2.5   
1   540.0                 0.0      0.0  162.0               2.5   
2   332.5               142.5      0.0  228.0               0.0   
3   332.5               142.5      0.0  228.0               0.0   
4   198.6               132.4      0.0  192.0               0.0   

   Coarse Aggregate  Fine Aggregate  Age  Strength  
0            1040.0           676.0   28     79.99  
1            1055.0           676.0   28     61.89  
2             932.0           594.0  270     40.27  
3             932.0           594.0  365     41.05  
4             978.4           825.5  360     44.30

In [10]:
concrete_norm = concrete.copy()
concrete_norm = apply_data_normalisation(concrete, concrete.columns)
concrete_norm.head()

C:\Users\mian_\AppData\Local\Temp\ipykernel_5856\3235113012.py:15: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise in a future error of pandas. Value '[-0.27959729 -0.27959729  3.55134048 ... -0.27959729 -0.27959729
 -0.27959729]' has dtype incompatible with int64, please explicitly cast to a compatible dtype first.
  norm_data.loc[:,feature] = feat_norm


Cement  Blast Furnace Slag   Fly Ash     Water  Superplasticizer  \
0  2.476712           -0.856472 -0.846733 -0.916319         -0.620147   
1  2.476712           -0.856472 -0.846733 -0.916319         -0.620147   
2  0.491187            0.795140 -0.846733  2.174405         -1.038638   
3  0.491187            0.795140 -0.846733  2.174405         -1.038638   
4 -0.790075            0.678079 -0.846733  0.488555         -1.038638   

   Coarse Aggregate  Fine Aggregate       Age  Strength  
0          0.862735       -1.217079 -0.279597  2.644123  
1          1.055651       -1.217079 -0.279597  1.560663  
2         -0.526262       -2.239829  3.551340  0.266498  
3         -0.526262       -2.239829  5.055221  0.313188  
4          0.070492        0.647569  4.976069  0.507732

In [11]:
# train data
concreteX = concrete_norm.drop(columns=["Strength"], axis=1)
concreteY = concrete_norm["Strength"]

conc_xtrain, conc_xtest, conc_ytrain, conc_ytest = train_test_split(concreteX, 
                                                                    concreteY, 
                                                                    test_size=0.3,)


In [12]:
# train the model
# retrieve all mse values + store into a mse_values
model_history = model.fit(conc_xtrain, conc_ytrain, validation_data=(conc_xtest, conc_ytest), epochs=50)
mse_values = model_history.history["val_loss"]

# predict the model
conc_pred = model.predict(conc_xtest)

# evaluate the model + create list of mse
mse = mean_squared_error(conc_ytest, conc_pred)

Epoch 1/50
23/23 ━━━━━━━━━━━━━━━━━━━━ 5s 36ms/step - loss: 1.1671 - val_loss: 1.0405
Epoch 2/50
23/23 ━━━━━━━━━━━━━━━━━━━━ 0s 10ms/step - loss: 0.9359 - val_loss: 0.9343
Epoch 3/50
23/23 ━━━━━━━━━━━━━━━━━━━━ 0s 9ms/step - loss: 0.9151 - val_loss: 0.8422
Epoch 4/50
23/23 ━━━━━━━━━━━━━━━━━━━━ 0s 10ms/step - loss: 0.8033 - val_loss: 0.7731
Epoch 5/50
23/23 ━━━━━━━━━━━━━━━━━━━━ 0s 9ms/step - loss: 0.7190 - val_loss: 0.7072
Epoch 6/50
23/23 ━━━━━━━━━━━━━━━━━━━━ 0s 9ms/step - loss: 0.6335 - val_loss: 0.6505
Epoch 7/50
23/23 ━━━━━━━━━━━━━━━━━━━━ 0s 10ms/step - loss: 0.6104 - val_loss: 0.6031
Epoch 8/50
23/23 ━━━━━━━━━━━━━━━━━━━━ 0s 10ms/step - loss: 0.5532 - val_loss: 0.5575
Epoch 9/50
23/23 ━━━━━━━━━━━━━━━━━━━━ 0s 9ms/step - loss: 0.5361 - val_loss: 0.5181
Epoch 10/50
23/23 ━━━━━━━━━━━━━━━━━━━━ 0s 9ms/step - loss: 0.5392 - val_loss: 0.4855
Epoch 11/50
23/23 ━━━━━━━━━━━━━━━━━━━━ 0s 10ms/step - loss: 0.4998 - val_loss: 0.4617
Epoch 12/50
23/23 ━━━━━━━━━━━━━━━━━━━━ 0s 10ms/step - loss: 0.5184 -

In [13]:
# mean squared error: result
print(f"mean squared error at (50/50): {mse}")

mean squared error at (50/50): 0.17562226994467575


In [14]:
# mse values
mse_list = mse_values
epoch_list = [e for e in range(1,len(mse_list)+1)]

print("Epoch - MSE value")
for ep, mse_val in zip(epoch_list, mse_list):
  print(f"{ep} - {mse_val}")

Epoch - MSE value
1 - 1.0404953956604004
2 - 0.934345543384552
3 - 0.8421595096588135
4 - 0.773057758808136
5 - 0.7072018384933472
6 - 0.6504928469657898
7 - 0.603073239326477
8 - 0.5574591755867004
9 - 0.5180734395980835
10 - 0.48549482226371765
11 - 0.461669385433197
12 - 0.4361037611961365
13 - 0.41606074571609497
14 - 0.3937752842903137
15 - 0.3793351650238037
16 - 0.35519543290138245
17 - 0.33948007225990295
18 - 0.3193742632865906
19 - 0.30191364884376526
20 - 0.28536829352378845
21 - 0.2821349501609802
22 - 0.269659161567688
23 - 0.25828230381011963
24 - 0.25162699818611145
25 - 0.24480922520160675
26 - 0.23763830959796906
27 - 0.23799945414066315
28 - 0.22925925254821777
29 - 0.22143490612506866
30 - 0.22551749646663666
31 - 0.21265952289104462
32 - 0.20888113975524902
33 - 0.20421920716762543
34 - 0.20706389844417572
35 - 0.20363856852054596
36 - 0.20257239043712616
37 - 0.19717846810817719
38 - 0.2087109535932541
39 - 0.19833315908908844
40 - 0.19723255932331085
41 - 0.186739

In [15]:
# mse values: mean and std
# convert epoch and mse into dataframe
# take the mean and std of it

def create_mse_dict(epoch: list, mse_val:list) -> dict:
  # create a dict for all mse values
  mse_dict = {}
  mse_sample = [epoch, mse_val]
  mse_feat = ["Epoch","MSE"]

  for feat, sample in zip(mse_feat, mse_sample):
    mse_dict[feat] = sample
  return mse_dict

def convert_mse_data(epoch, mse_val):
  mse_dict = create_mse_dict(epoch, mse_val)
  mse_data = pd.DataFrame(mse_dict)
  return mse_data

In [16]:
mse_dict_ = create_mse_dict(epoch_list, mse_list)
mse_data = convert_mse_data(epoch_list, mse_list)
mse_data.head()

Epoch       MSE
0      1  1.040495
1      2  0.934346
2      3  0.842160
3      4  0.773058
4      5  0.707202

In [17]:
# find the statistical measurement
mse_mean = mse_data["MSE"].mean()
mse_std = mse_data["MSE"].std()
print(f"Mean value of MSE: {mse_mean:.3f}")
print(f"Standard Deviation of MSE: {mse_std:.3f}")

Mean value of MSE: 0.343
Standard Deviation of MSE: 0.214


**How does the mean of the mean squared errors compare to that from Step A?**

+ mean of mse from part B (with normalisation): 0.589
+ mean of mse from part D (with normalisation): 0.343

By increasing the number of hidden layers, the mean value of the mean squared error decreases. By adding 3 layers in the model which originally had one hidden layer, the mean vanlue of mse in part B reduced from 0.589 to 0.343 in part D